### ISSUE TITLE
IMAGE CAPTIONING

### ISSUE THREAD

**Issue by daksh01125** [daksh01125/TIc-Tac-Toe-AI#4]
def caption_image_beam_search(encoder, decoder, image, word_map, beam_size=3, device='cpu'):
    k = beam_size
    vocab_size = len(word_map)

    encoder.eval()
    decoder.eval()

    image = image.to(device)
    encoder_out = encoder(image.unsqueeze(0))  # (1, num_pixels, encoder_dim)
    encoder_dim = encoder_out.size(-1)
    num_pixels = encoder_out.size(1)

    # Flatten encoding
    encoder_out = encoder_out.view(1, -1, encoder_dim)

    # We'll treat the problem as having a batch size of k
    encoder_out = encoder_out.expand(k, num_pixels, encoder_dim)

    # Tensor to store top k sequences; now they contain only the start token
    start_token = word_map['<start>']
    sequences = torch.full((k, 1), start_token, dtype=torch.long).to(device)

    # Scores for each sequence
    top_k_scores = torch.zeros(k, 1).to(device)

    # Lists to store completed sequences and scores
    complete_seqs = []
    complete_seqs_scores = []

    # Initialize hidden and cell states
    h, c = decoder.init_hidden_state(encoder_out)

    # Start decoding
    step = 1
    while True:
        embeddings = decoder.embedding(sequences[:, -1])  # (k, embed_dim)
        awe, alpha = decoder.attention(encoder_out, h)  # (k, encoder_dim), (k, num_pixels)
        gate = decoder.sigmoid(decoder.f_beta(h))
        awe = gate * awe

        lstm_input = torch.cat([embeddings, awe], dim=1)
        h, c = decoder.decode_step(lstm_input, (h, c))
        scores = decoder.fc(h)  # (k, vocab_size)
        scores = F.log_softmax(scores, dim=1)

        # Add scores to previous scores
        scores = top_k_scores.expand_as(scores) + scores

        # Flatten for topk
        if step == 1:
            top_k_scores, top_k_words = scores[0].topk(k, 0, True, True)
        else:
            top_k_scores, top_k_words = scores.view(-1).topk(k, 0, True, True)

        # Convert flattened indices to actual indices
        prev_word_inds = top_k_words // vocab_size
        next_word_inds = top_k_words % vocab_size

        # Append new words to sequences
        sequences = torch.cat([sequences[prev_word_inds], next_word_inds.unsqueeze(1)], dim=1)

        # Check for complete sequences
        incomplete_inds = []
        complete_inds = []

        for i, next_word in enumerate(next_word_inds):
            if next_word == word_map['<end>']:
                complete_seqs.append(sequences[i].tolist())
                complete_seqs_scores.append(top_k_scores[i])
                complete_inds.append(i)
            else:
                incomplete_inds.append(i)

        # Reduce beam size
        if len(complete_inds) > 0:
            k -= len(complete_inds)
            if k == 0:
                break
            sequences = sequences[incomplete_inds]
            h = h[prev_word_inds[incomplete_inds]]
            c = c[prev_word_inds[incomplete_inds]]
            encoder_out = encoder_out[prev_word_inds[incomplete_inds]]
            top_k_scores = top_k_scores[incomplete_inds].unsqueeze(1)

        # Stop if max length reached
        if step > 50:
            break
        step += 1

    if len(complete_seqs_scores) == 0:
        complete_seqs.append(sequences[0].tolist())
        complete_seqs_scores.append(top_k_scores[0])

    i = complete_seqs_scores.index(max(complete_seqs_scores))
    seq = complete_seqs[i]

    return seq
